model/vm_pool.rb

```diff
@@ -5,21 +5,22 @@
 class VmPool < Sequel::Model
   one_to_one :strand, key: :id
   one_to_many :vms, key: :pool_id
+  one_to_many :idle_vms, key: :pool_id, class: Vm, conditions: {tainted_at: nil, display_state: "running"}
 
   include ResourceMethods
 
   include SemaphoreMethods
   semaphore :destroy
 
-  def pick_vm
-    DB.transaction do
-      # first lock the whole pool in the join table so that no other thread can
-      # pick a vm from this pool
-      vms_dataset.for_update.all
-      pick_vm_id_q = vms_dataset.left_join(:github_runner, vm_id: :id)
-        .where(Sequel[:github_runner][:vm_id] => nil, Sequel[:vm][:display_state] => "running")
-        .select(Sequel[:vm][:id])
-      Vm.where(id: pick_vm_id_q).first
-    end
+  def pick_vm(new_name)
+    vm_hash = DB[:vm]
+      .returning
+      .with(:candidate, idle_vms_dataset.limit(1))
+      .from(:vm, :candidate)
+      .where(Sequel[:vm][:id] => Sequel[:candidate][:id])
+      .update(name: new_name, tainted_at: Time.now)
+      .first
```

Comment:
Personally, I am having difficulty to visualize what kind of SQL query this would generate when `returning`, `with`, `update` and `first` are combined, hence I'm not confident on its correctness. It feels like if we plainly write the SQL query here, it would be easier to understand and reason about it.

BTW, do we have a convention about whether we prefer Sequel or SQL for non-trivial stuff? Personally I use both without much thinking, so I don't have a strong opinion.

Reply:
Personally I use either. Evans, for his part, always uses the DSL. 

There are some interesting composition problems with using both that I just not thought about. One particularly ugly one is how datasets with raw SQL will silently accept, but not do anything with, additional dataset methods. There's a whole plugin for this:

https://sequel.jeremyevans.net/rdoc-plugins/files/lib/sequel/extensions/implicit_subquery_rb.html

I thought about applying it, but decided the bloating was not worth it...at the time.

Consider this:

```
[4] clover-development(main)> DB["select * from pg_class"].where { oid > 7 }.sql
=> "select * from pg_class"
```

I'd actually want a plugin that errored whenever such a ignored composition was done...I will write that into the mailing list.